xarray/plot/dataset_plot.py

```diff
@@ -0,0 +1,393 @@
+        The mapping from data values to color space. If not provided, this
+        will be either be ``viridis`` (if the function infers a sequential
+        dataset) or ``RdBu_r`` (if the function infers a diverging dataset).
+        When `Seaborn` is installed, ``cmap`` may also be a `seaborn`
+        color palette. If ``cmap`` is seaborn color palette and the plot type
+        is not ``contour`` or ``contourf``, ``levels`` must also be specified.
+    colors : discrete colors to plot, optional
+        A single color or a list of colors. If the plot type is not ``contour``
+        or ``contourf``, the ``levels`` argument is required.
+    center : float, optional
+        The value at which to center the colormap. Passing this value implies
+        use of a diverging colormap. Setting it to ``False`` prevents use of a
+        diverging colormap.
+    robust : bool, optional
+        If True and ``vmin`` or ``vmax`` are absent, the colormap range is
+        computed with 2nd and 98th percentiles instead of the extreme values.
+    extend : {'neither', 'both', 'min', 'max'}, optional
+        How to draw arrows extending the colorbar beyond its limits. If not
+        provided, extend is inferred from vmin, vmax and the data limits.
+    levels : int or list-like object, optional
+        Split the colormap (cmap) into discrete color intervals. If an integer
+        is provided, "nice" levels are chosen based on the data range: this can
+        imply that the final number of levels is not exactly the expected one.
+        Setting ``vmin`` and/or ``vmax`` with ``levels=N`` is equivalent to
+        setting ``levels=np.linspace(vmin, vmax, N)``.
+    **kwargs : optional
+        Additional keyword arguments to matplotlib
+    """
+
+    # Build on the original docstring
+    plotfunc.__doc__ = '%s\n%s' % (plotfunc.__doc__, commondoc)
+
+    @functools.wraps(plotfunc)
+    def newplotfunc(ds, x=None, y=None, hue=None, hue_style=None,
+                    col=None, row=None, ax=None, figsize=None, size=None,
+                    col_wrap=None, sharex=True, sharey=True, aspect=None,
+                    subplot_kws=None, add_guide=None, cbar_kwargs=None,
+                    cbar_ax=None, vmin=None, vmax=None,
+                    norm=None, infer_intervals=None, center=None, levels=None,
+                    robust=None, colors=None, extend=None, cmap=None,
+                    **kwargs):
+
+        _is_facetgrid = kwargs.pop('_is_facetgrid', False)
+        if _is_facetgrid:  # facetgrid call
+            meta_data = kwargs.pop('meta_data')
+        else:
+            meta_data = _infer_meta_data(ds, x, y, hue, hue_style, add_guide)
+
+        hue_style = meta_data['hue_style']
+
+        # handle facetgrids first
+        if col or row:
+            allargs = locals().copy()
+            allargs['plotfunc'] = globals()[plotfunc.__name__]
+            allargs['data'] = ds
+            # TODO dcherian: why do I need to remove kwargs?
```

Comment:
I feel like there is probably a better design for these methods than use `locals()`, but I'm not entirely sure what. This is probably an indication that our functions have too many arguments!

Reply:
:)